src/backend/src/signer.rs

```diff
@@ -0,0 +1,60 @@
+//! Code for inetracting with the chain fusion signer.
+use crate::state::{CYCLES_LEDGER, SIGNER};
+use candid::{CandidType, Deserialize, Nat, Principal};
+use ic_cycles_ledger_client::{Account, ApproveArgs, ApproveError, Service as CyclesLedgerService};
+use ic_ledger_types::Subaccount;
+use serde_bytes::ByteBuf;
+
+#[derive(CandidType, Deserialize, Debug, Clone, Eq, PartialEq)]
+pub enum AllowSigningError {
+    Other(String),
+    FailedToContactCyclesLedger,
+    ApproveError(ApproveError),
+}
+
+/// Enables the user to sign transactions.
+///
+/// Signing costs cycles.  Managing that cycle payment can be painful so we take care of that.
+pub async fn allow_signing() -> Result<(), AllowSigningError> {
+    let cycles_ledger: Principal = *CYCLES_LEDGER;
+    let signer: Principal = *SIGNER;
+    let caller = ic_cdk::caller();
+    let expected_fee = 1_000_000_000u64;
+    let amount = Nat::from(100 * expected_fee);
+    CyclesLedgerService(cycles_ledger)
+        .icrc_2_approve(&ApproveArgs {
+            spender: Account {
+                owner: signer,
+                subaccount: Some(principal2account(&caller)),
+            },
+            amount,
+            created_at_time: None,
+            expected_allowance: None,
+            expires_at: None,
+            fee: None,
+            from_subaccount: None,
+            memo: None,
+        })
+        .await
+        .map_err(|_| AllowSigningError::FailedToContactCyclesLedger)?
+        .0
+        .map_err(AllowSigningError::ApproveError)?;
+    Ok(())
+}
+
+const SUB_ACCOUNT_ZERO: Subaccount = Subaccount([0; 32]);
+#[must_use]
+pub fn principal2account(principal: &Principal) -> ByteBuf {
```

Comment:
Why is it public?

Reply:
If it is public, the editor will make it obvious if there are two copies of the same function in the codebase.

Note: I merged a PR upstream that provides an account identifier as bytes directly.  There is a Jira ticket to replace all of these `principal2account()` calls.  That can be done after the next ic-cdk release.

Reply:
Thanks for the explanation